## Software Demo: Branching in SageMath (and LiE)

Let $G$ be a compact, simply connected, **semisimple** Lie group. With SageMath we can study its (complex) representations.

(Equivalently, representations of compact/complex semisimple Lie algebras.)

We let $\mathfrak{t}\subset\mathfrak{g}$ be a Cartan subalgebra of $\mathfrak{g}=\mathrm{Lie}(G)$. The **roots** of $\mathfrak{g}$ are elements of $\alpha\in\mathfrak{t}^\ast\setminus\{0\}$ which satisfy
$$\forall X\in\mathfrak{t},\qquad [X,X_\alpha]=\mathrm{i}\alpha(X)X_\alpha$$
for some **root vector** $X_\alpha\in\mathfrak{g}^\mathbb{C}$.

More generally, for any representation $\pi: G\to\mathrm{GL}(V)$, a **weight** is an element $\lambda\in\mathfrak{t}^\ast$ such that
$$\forall X\in\mathfrak{t},\qquad \pi_\ast(X)v_\lambda=\mathrm{i}\lambda(X)v_\lambda$$
for some **weight vector** $v_\lambda\in V$.

### 1) Playing around with root systems

$G$ has a **Cartan type** corresponding to its system of roots. For example:

* $B_4$ corresponds to the group $\mathrm{Spin}(9)$, or the Lie algebra $\mathfrak{so}(9)$.

In [63]:
R=RootSystem("B4")
R.dynkin_diagram()

O---O---O=>=O
1   2   3   4   
B4

The roots are available as vectors (in some ambient space containing $\mathfrak{t}^\ast$). Here are the simple roots for $B_4$.

In [64]:
R.ambient_space().simple_roots()

Finite family {1: (1, -1, 0, 0), 2: (0, 1, -1, 0), 3: (0, 0, 1, -1), 4: (0, 0, 0, 1)}

The *fundamental weights* of $G$ are a $\mathbb{Z}_{\geq0}$-basis for the set of dominant integral weights, which parametrize the irreducible representations of $G$.

In [65]:
R.ambient_space().fundamental_weights()

Finite family {1: (1, 0, 0, 0), 2: (1, 1, 0, 0), 3: (1, 1, 1, 0), 4: (1/2, 1/2, 1/2, 1/2)}

* The ambient space and its basis follows the convention of Bourbaki. For example for $A_n$, corresponding to $G=\mathrm{SU}(n+1)$, we think of $\mathfrak{t}\subset\mathfrak{su}(n+1)$ as the space $\{x\in\mathbb{R}^{n+1}\,|\,x_1+\ldots x_{n+1}=0\}$.

In [66]:
R=RootSystem("A4")
R.ambient_space().simple_roots()

Finite family {1: (1, -1, 0, 0, 0), 2: (0, 1, -1, 0, 0), 3: (0, 0, 1, -1, 0), 4: (0, 0, 0, 1, -1)}

SageMath can only handle **semisimple** groups, with one exception: Results for $A_n$, i.e. $G=\mathrm{SU}(n+1)$, or $\mathrm{SL}(n+1,\mathbb{C})$ can also be interpreted as results for $\mathrm{U}(n+1)$ resp. $\mathrm{GL}(n+1,\mathbb{C})$.

In [67]:
fw=R.ambient_space().fundamental_weights()
fw

Finite family {1: (1, 0, 0, 0, 0), 2: (1, 1, 0, 0, 0), 3: (1, 1, 1, 0, 0), 4: (1, 1, 1, 1, 0)}

In [68]:
[w.coerce_to_sl() for w in fw]

[(4/5, -1/5, -1/5, -1/5, -1/5),
 (3/5, 3/5, -2/5, -2/5, -2/5),
 (2/5, 2/5, 2/5, -3/5, -3/5),
 (1/5, 1/5, 1/5, 1/5, -4/5)]

* $G_2$ is an exceptional root system/Lie group. We can plot its roots:

In [69]:
posrootsG2=RootSystem("G2").ambient_space().positive_roots()
posrootsG2

[(0, 1, -1), (1, -2, 1), (1, -1, 0), (1, 0, -1), (1, 1, -2), (2, -1, -1)]

In [70]:
p=sum([plot(vector(v),thickness=2)+plot(-vector(v),thickness=2) for v in posrootsG2])
p.show(viewpoint=[[-0.1875,-0.4511,-0.8726],140.14])

Graphics3d Object

* Reducible root systems are also no problem.

In [71]:
R=RootSystem("A1xA1xB2")
R.dynkin_diagram()

O
1   
O
2   
O=>=O
3   4   
A1xA1xB2

In [72]:
R.ambient_space().simple_roots()

Finite family {1: (1, -1, 0, 0, 0, 0), 2: (0, 0, 1, -1, 0, 0), 3: (0, 0, 0, 0, 1, -1), 4: (0, 0, 0, 0, 0, 1)}

### 2) Representations and their weights

SageMath can calculate with the **character ring** of $G$, which is generated by its finite-dimensional irreducible representations. These are labeled by their highest weights.

The option $\texttt{style="coroots"}$ ensures that weights are expressed in the basis of fundamental weights.

In [73]:
G=WeylCharacterRing("B3",style="coroots"); G

The Weyl Character Ring of Type B3 with Integer Ring coefficients

The standard representation of $\mathfrak{so}(7)$ is $\mathbb{C}^7$ and corresponds to the first fundamental weight.

In [76]:
G(0,1,0).degree()

21

In [77]:
G(0,1,0).highest_weight()

(1, 1, 0)

In [78]:
G(1,0,0).weight_multiplicities()

{(1, 0, 0): 1,
 (0, 1, 0): 1,
 (0, 0, 1): 1,
 (0, 0, 0): 1,
 (0, 0, -1): 1,
 (0, -1, 0): 1,
 (-1, 0, 0): 1}

Here is the adjoint representation of $\mathfrak{so}(7)$.

In [79]:
G.adjoint_representation()

B3(0,1,0)

The product in the character ring corresponds precisely to the tensor product, and SageMath automatically decomposes it into irreducible pieces. Here we see $\mathbb{C}^7\otimes\mathbb{C}^7$, and $\mathbb{C}^7\otimes\mathbb{C}^7\otimes\mathbb{C}^7$.

In [80]:
G(1,0,0)*G(1,0,0)

B3(0,0,0) + B3(0,1,0) + B3(2,0,0)

In [81]:
G(1,0,0)^3

3*B3(1,0,0) + B3(0,0,2) + 2*B3(1,1,0) + B3(3,0,0)

Of course there are also symmetric powers and exterior powers.

In [82]:
G(1,0,0).symmetric_power(2)

B3(0,0,0) + B3(2,0,0)

In [84]:
G(1,0,0).exterior_power(8)

0

We can count the multiplicity $\dim\mathrm{Hom}_G(V,W)$ of an irreducible representation $V$ in a representation $W$.

The multiplicity of $\Lambda^2\mathbb{C}^7$ in $\mathbb{C}^7\otimes\mathbb{C}^7$ is 1:

In [85]:
rep=G(1,0,0)*G(1,0,0)
rep.multiplicity(G(0,1,0))

1

SageMath also support the **Weyl group** of $G$, acting as a matrix group on the ambient space. But we will not do anything with it.

In [88]:
W=G.space().weyl_group(); W

Weyl Group of type ['B', 3] (as a matrix group acting on the ambient space)

In [87]:
W.gens()

(
[0 1 0]  [1 0 0]  [ 1  0  0]
[1 0 0]  [0 0 1]  [ 0  1  0]
[0 0 1], [0 1 0], [ 0  0 -1]
)

Let us calculate the dimensions of the spinor representations for the odd spin groups $B_n$.

In [89]:
for n in [1..8]:
    GB=WeylCharacterRing(["B",n],style="coroots")
    spin=GB(GB.fundamental_weights()[n])
    print("Spin({}): dim(spinor module) = {}".format(2*n+1,spin.degree()))

Spin(3): dim(spinor module) = 2
Spin(5): dim(spinor module) = 4
Spin(7): dim(spinor module) = 8
Spin(9): dim(spinor module) = 16
Spin(11): dim(spinor module) = 32
Spin(13): dim(spinor module) = 64
Spin(15): dim(spinor module) = 128
Spin(17): dim(spinor module) = 256


### 3) Casimir constants

Choose any $G$-invariant inner product on the Lie algebra $\mathfrak{g}$, and let $(X_i)$ be an orthonormal basis of $\mathfrak{g}$.

For any representation $\pi: G\to\mathrm{GL}(V)$, its **Casimir operator** is defined by
$$\mathrm{Cas}^G_\pi=-\sum_i\pi_\ast(X_i)^2.$$
By Schur's Lemma, $\mathrm{Cas}^G_\pi$ acts as multiplication by a constant if $\pi$ is irreducible. **Freudenthal's formula** states that the Casimir constant on the irreducible representation with highest weight $\lambda\in\mathfrak{t}^\ast$ is
$$\mathrm{Cas}^G_\lambda=\langle\lambda,\lambda+2\rho\rangle,$$
where $\rho\in\mathfrak{t}^\ast$ is the half-sum of positive roots.

In [90]:
G.space().rho()

(5/2, 3/2, 1/2)

In [91]:
G.space().rho()==1/2*sum(G.space().positive_roots())

True

In [92]:
def Casimir(rep):
    hweight=rep.highest_weight()
    delta=rep.parent().space().rho()
    return hweight.inner_product(hweight+2*delta)

In [93]:
[Casimir(G(G.fundamental_weights()[n])) for n in [1..3]]

[6, 10, 21/4]

The value depends of course on the chosen inner product. This can cause a lot of headache in practice; one remedy is the fact that the Casimir operator defined with respect to **minus the Killing form** assumes the value 1 on the adjoint representation.

In [94]:
CasAd=Casimir(G.adjoint_representation())
def CasimirKilling(rep):
    return Casimir(rep)/CasAd

In [95]:
[CasimirKilling(G(G.fundamental_weights()[n])) for n in [1..3]]

[3/5, 1, 21/40]

### 4) Branching rules

Let $G$ be compact (simply connected) and semisimple, and let $H\subset G$ be a closed subgroup. Then $H$ is a **reductive** Lie group: it can be written as $H=H_{ss}\times T$, where $H_{ss}$ is semisimple and $T$ is a torus (which is central in $H$).

For now, we assume $H=H_{ss}$.

If $\pi: G\to\mathrm{GL}(V)$ is a representation, we can restrict (**branch**) it to a representation $\pi|_H$ of $H$. The **branching rule** for the pair tells us how a Cartan subalgebra $\mathfrak{t}_{\mathfrak{h}}$ of $\mathfrak{h}$ embeds into a Cartan subalgebra $\mathfrak{t}_{\mathfrak{g}}$ of $\mathfrak{g}$; or conversely, how weights of $\mathfrak{g}$ restrict to weights of $\mathfrak{h}$.

Let us consider the pair of subalgebras $H=\mathrm{Spin}(7)\subset G=\mathrm{Spin}(8)$. These define the symmetric space $G/H=S^7$.

In [96]:
G=WeylCharacterRing("D4",style="coroots")
H=WeylCharacterRing("B3",style="coroots")
b=branching_rule("D4","B3",rule="symmetric"); b

symmetric branching rule D4 => B3

Now, let us see the branching in action. This corresponds to $\mathbb{C}^8=\mathbb{C}\oplus\mathbb{C}^7$:

In [97]:
b.branch(G(1,0,0,0))

B3(0,0,0) + B3(1,0,0)

And here we find the decomposition $\mathfrak{g}^\mathbb{C}\cong\mathbb{C}^7\oplus\mathfrak{h}^\mathbb{C}$:

In [98]:
b.branch(G.adjoint_representation())

B3(1,0,0) + B3(0,1,0)

SageMath has many ways to generate branching rules. For $G$ simple and $\mathrm{rank}(G)\leq8$, it knows how the maximal subgroups (resp. their semisimple parts) are embedded.

In [102]:
G.maximal_subgroups()

B3:branching_rule("D4","B3","symmetric")
A2:branching_rule("D4","A2(1,1)","plethysm")
A1xC2:branching_rule("D4","C1xC2","tensor")*branching_rule("C1xC2","A1xC2",[branching_rule("C1","A1","isomorphic"),"identity"])
A1xA1xA1xA1:branching_rule("D4","D2xD2","orthogonal_sum")*branching_rule("D2xD2","A1xA1xA1xA1",[branching_rule("D2","A1xA1","isomorphic"),branching_rule("D2","A1xA1","isomorphic")])


In [103]:
WeylCharacterRing("A6").maximal_subgroups()

A5:branching_rule("A6","A5","levi")
B3:branching_rule("A6","B3","symmetric")
A1xA4:branching_rule("A6","A1xA4","levi")
A2xA3:branching_rule("A6","A2xA3","levi")


In [104]:
WeylCharacterRing("E6").maximal_subgroups()

D5:branching_rule("E6","D5","levi")
C4:branching_rule("E6","C4","symmetric")
F4:branching_rule("E6","F4","symmetric")
A2:branching_rule("E6","A2","miscellaneous")
G2:branching_rule("E6","G2","miscellaneous")
A2xG2:branching_rule("E6","A2xG2","miscellaneous")
A1xA5:branching_rule("E6","A1xA5","extended")
A2xA2xA2:branching_rule("E6","A2xA2xA2","extended")


In [105]:
WeylCharacterRing("F4").maximal_subgroups()

B4:branching_rule("F4","B4","extended")
A1:branching_rule("F4","A1","ii")
A1xG2:branching_rule("F4","A1xG2","miscellaneous")
A1xC3:branching_rule("F4","A1xC3","extended")
A2xA2:branching_rule("F4","A2xA2","extended")


In [106]:
branching_rule("E6","F4","symmetric")*branching_rule("F4","B4","extended")

composite branching rule E6 => (symmetric) F4 => (extended) B4

We can peer into what the branching rule does to roots and weights.

In [107]:
b.describe(verbose=True)


    O 4
    |
    |
O---O---O
1   |2  3   
    |
    O 0
D4~
root restrictions D4 => B3:

O---O=>=O
1   2   3   
B3

0 => root (-1, -1, 0)
1 => 1
2 => 2
3 => 3
4 => 3

fundamental weight restrictions D4 => B3:
1 => (1, 0, 0)
2 => (0, 1, 0)
3 => (0, 0, 1)
4 => (0, 0, 1)


For branching to non-maximal subgroups, one may compose branching rules. Let us take a look at the four different conjugacy classes of $A_1$ subgroups inside $G_2$.

In [108]:
G2=WeylCharacterRing("G2",style="coroots")
G2.maximal_subgroups()

A2:branching_rule("G2","A2","extended")
A1:branching_rule("G2","A1","i")
A1xA1:branching_rule("G2","A1xA1","extended")


In [109]:
bMax=branching_rule("G2","A1","i"); bMax
bLeft=branching_rule("G2","A1xA1","extended")*branching_rule("A1xA1","A1","proj1")
bRight=branching_rule("G2","A1xA1","extended")*branching_rule("A1xA1","A1","proj2")
bDiag=branching_rule("G2","A1xA1","extended")*branching_rule("A1xA1","A1","diagonal")
[bMax,bLeft,bRight,bDiag]

[i branching rule G2 => A1,
 composite branching rule G2 => (extended) A1xA1 => (proj1) A1,
 composite branching rule G2 => (extended) A1xA1 => (proj2) A1,
 composite branching rule G2 => (extended) A1xA1 => (diagonal) A1]

Let us try to branch some representations of $G_2$ to compare the branching rules.

Recall that the standard representation of $G_2$ is $\mathbb{C}^7=\mathrm{Im}(\mathbb{O})^\mathbb{C}$. The decompositions below all have an interpretation in terms of the action on octonions.

In [110]:
A1=WeylCharacterRing("A1",style="coroots")
[G2(1,0).branch(A1,b) for b in [bMax,bLeft,bRight,bDiag]]

[A1(6), 2*A1(1) + A1(2), 3*A1(0) + 2*A1(1), A1(0) + 2*A1(2)]

A particularly powerful one is the $\texttt{plethysm}$ rule. It works if $G$ is a classical group ($\mathrm{SU}(n)$, $\mathrm{SO}(n)$, $\mathrm{Sp}(n/2)$, and $H$ is embedded into $G$ by a representation $\pi: H\to\mathrm{GL}(n,\mathbb{C})$.

Let's test this: we embed $\mathrm{Spin}(7)$ into $\mathrm{Spin}(\mathfrak{so}(7))$ via the (lift of the) adjoint representation.

In [111]:
ad=H.adjoint_representation(); [ad, ad.degree()]

[B3(0,1,0), 21]

$\dim\mathfrak{so}(7)=21$, so $\mathrm{Spin}(\mathfrak{so}(7))$ is of Cartan type $B_{10}$ because $21=2\cdot 10+1$.

In [112]:
b=branching_rule("B10","B3(0,1,0)",rule="plethysm"); b

plethysm (along B3(0,1,0)) branching rule B10 => B3

In [113]:
B10=WeylCharacterRing("B10",style="coroots")
b.branch(B10.adjoint_representation())

B3(0,1,0) + B3(1,0,2)

In [114]:
H(1,0,2).degree()

189

So $\mathfrak{so}(21)^\mathbb{C}\cong\mathfrak{so}(7)^\mathbb{C}\oplus[1,0,2]$, and $\mathrm{Spin}(21)/\mathrm{Spin}(7)$ is an 189-dimensional isotropy-irreducible space!

### 5) Root vectors

SageMath also allows computations inside a given Lie algebra. For semisimple Lie algebras, we have access to the root vectors.

In [115]:
g2=LieAlgebra(QQ, cartan_type="G2"); g2

Lie algebra of ['G', 2] in the Chevalley basis

Let $\mathfrak{g}$ be a semisimple Lie algebra with root space decomposition
$$\mathfrak{g}^\mathbb{C}=\mathfrak{t}^\mathbb{C}\oplus\bigoplus_{\alpha}\mathfrak{g}_\alpha,$$
where $\alpha$ runs over all roots and $\mathfrak{g}_\alpha$ is the simultaneous eigenspace to the root $\alpha$.

A **Chevalley basis** is a basis adapted to this root space decomposition: to each root $\alpha$, it associates a vector $E_\alpha\in\mathfrak{g}_\alpha$, and to each _simple_ root $\alpha$ a vector $H_\alpha\in\mathfrak{t}$.

In this part of SageMath, roots are abstractly represented as linear combinations of simple roots. But we can also convert them to the "ambient space" we saw earlier.

In [116]:
R=RootSystem("G2")
roots=R.root_space().roots()
roots

[alpha[1],
 alpha[2],
 alpha[1] + alpha[2],
 3*alpha[1] + alpha[2],
 2*alpha[1] + alpha[2],
 3*alpha[1] + 2*alpha[2],
 -alpha[1],
 -alpha[2],
 -alpha[1] - alpha[2],
 -3*alpha[1] - alpha[2],
 -2*alpha[1] - alpha[2],
 -3*alpha[1] - 2*alpha[2]]

In [117]:
f=R.root_space().to_ambient_space_morphism()
[f(r) for r in roots]

[(0, 1, -1),
 (1, -2, 1),
 (1, -1, 0),
 (1, 1, -2),
 (1, 0, -1),
 (2, -1, -1),
 (0, -1, 1),
 (-1, 2, -1),
 (-1, 1, 0),
 (-1, -1, 2),
 (-1, 0, 1),
 (-2, 1, 1)]

The Chevalley basis in SageMath consists also of abstract vectors, and is indexed by the roots (and simple coroots) above.

In [118]:
B=basis(g2)
alpha=R.root_space().simple_roots()
alphacheck=R.root_space().simple_coroots()
B

Finite family {alpha[2]: E[alpha[2]], alpha[1]: E[alpha[1]], alpha[1] + alpha[2]: E[alpha[1] + alpha[2]], 2*alpha[1] + alpha[2]: E[2*alpha[1] + alpha[2]], 3*alpha[1] + alpha[2]: E[3*alpha[1] + alpha[2]], 3*alpha[1] + 2*alpha[2]: E[3*alpha[1] + 2*alpha[2]], alphacheck[1]: h1, alphacheck[2]: h2, -alpha[2]: E[-alpha[2]], -alpha[1]: E[-alpha[1]], -alpha[1] - alpha[2]: E[-alpha[1] - alpha[2]], -2*alpha[1] - alpha[2]: E[-2*alpha[1] - alpha[2]], -3*alpha[1] - alpha[2]: E[-3*alpha[1] - alpha[2]], -3*alpha[1] - 2*alpha[2]: E[-3*alpha[1] - 2*alpha[2]]}

In [119]:
B[alpha[1]+alpha[2]]

E[alpha[1] + alpha[2]]

In [120]:
B[alphacheck[1]]

h1

With this setup, we can now take Lie brackets! Recall that $[\mathfrak{g}_\alpha,\mathfrak{g}_\beta]\subset\mathfrak{g}_{\alpha+\beta}$ for $\alpha+\beta\neq0$, and $[\mathfrak{g}_\alpha,\mathfrak{g}_{-\alpha}]\subset\mathfrak{t}$.

In [121]:
v=B[alpha[1]]
w=B[alpha[1]+alpha[2]]
v.bracket(w)

-2*E[2*alpha[1] + alpha[2]]

In [122]:
u=B[-alpha[1]-alpha[2]]
u.bracket(w)

-h1 - 3*h2

Let us test the calculation of the Casimir operator on the adjoint representation. Since the Chevalley basis is not orthonormal under the Killing form $B_\mathfrak{g}$, we use the formula
$$\mathrm{Cas}^G_\pi=\sum_{i,j}K^{ij}\pi_\ast(b_i)\pi_\ast(b_j),$$
where $(b_i)$ is our basis and $K=(K_{ij})=(B_\mathfrak{g}(b_i,b_j))$.

In [123]:
K=Matrix([[g2.killing_form(v,w) for v in B] for w in B]); K

[  0   0   0   0   0   0   0   0   8   0   0   0   0   0]
[  0   0   0   0   0   0   0   0   0  24   0   0   0   0]
[  0   0   0   0   0   0   0   0   0   0  24   0   0   0]
[  0   0   0   0   0   0   0   0   0   0   0  24   0   0]
[  0   0   0   0   0   0   0   0   0   0   0   0   8   0]
[  0   0   0   0   0   0   0   0   0   0   0   0   0   8]
[  0   0   0   0   0   0  48 -24   0   0   0   0   0   0]
[  0   0   0   0   0   0 -24  16   0   0   0   0   0   0]
[  8   0   0   0   0   0   0   0   0   0   0   0   0   0]
[  0  24   0   0   0   0   0   0   0   0   0   0   0   0]
[  0   0  24   0   0   0   0   0   0   0   0   0   0   0]
[  0   0   0  24   0   0   0   0   0   0   0   0   0   0]
[  0   0   0   0   8   0   0   0   0   0   0   0   0   0]
[  0   0   0   0   0   8   0   0   0   0   0   0   0   0]

In [124]:
Kinv=K^(-1)
Blist=list(B)
def CasimirG2ByBasis(x):
    s=0
    for i in range(0,14):
        for j in range(0,14):
            s+=Kinv[i,j]*x.bracket(Blist[i]).bracket(Blist[j])
    return s

In [125]:
for v in B:
    print("{} -> {}".format(v,CasimirG2ByBasis(v)))

E[alpha[2]] -> E[alpha[2]]
E[alpha[1]] -> E[alpha[1]]
E[alpha[1] + alpha[2]] -> E[alpha[1] + alpha[2]]
E[2*alpha[1] + alpha[2]] -> E[2*alpha[1] + alpha[2]]
E[3*alpha[1] + alpha[2]] -> E[3*alpha[1] + alpha[2]]
E[3*alpha[1] + 2*alpha[2]] -> E[3*alpha[1] + 2*alpha[2]]
h1 -> h1
h2 -> h2
E[-alpha[2]] -> E[-alpha[2]]
E[-alpha[1]] -> E[-alpha[1]]
E[-alpha[1] - alpha[2]] -> E[-alpha[1] - alpha[2]]
E[-2*alpha[1] - alpha[2]] -> E[-2*alpha[1] - alpha[2]]
E[-3*alpha[1] - alpha[2]] -> E[-3*alpha[1] - alpha[2]]
E[-3*alpha[1] - 2*alpha[2]] -> E[-3*alpha[1] - 2*alpha[2]]


As we see, the Casimir operator really acts as the identity on the adjoint representation.

## LiE section

LiE allows a central torus, but only integer numbers. It is much faster in tensoring and branching than SageMath.

File reading command in LiE console: $\texttt{read file.lie}$

### 1) Example: Curvature invariants

$\texttt{curv.lie}$

Finding out where the invariants come from: scal, Ric, Weyl (fixed degree)

### 2) Restriction matrices

A **restriction matrix** is LiE's way of encoding branching rules. If $G$ and $H$ are reductive Lie groups, then the restriction matrix is the matrix of the restriction map $\mathfrak{t}_\mathfrak{g}^\ast\to\mathfrak{t}_\mathfrak{h}^\ast$ with respect to the bases of fundamental weights. (To be precise, its transpose, since LiE works with row vectors).

Example: LiE command $\texttt{max\_sub}$

Example: Branching for $\mathrm{S}(\mathrm{U}(p)\times\mathrm{U}(q))\subset\mathrm{SU}(p+q)$: $\texttt{ComplexGrassmannians.lie}$

Example: Branching for the $\mathrm{U}(2)$-subgroups of $G_2$: $\texttt{G2U2.lie}$

### 3) Example: Laplace eigenvalues

Calculation of Casimir constants: $\texttt{Casimir.lie}$

Eigenvalues of the Laplacian on functions for normal homogeneous spaces: $\texttt{FunctionLaplacian.lie}$

### 4) The Sage-LiE interface

LiE (like many other software packages) is accessible from SageMath!

In [126]:
LP=lie("1X[1,0]+2X[1,1]"); LP

1X[1,0] +2X[1,1]

In [127]:
P=LP.sage(); P

2*x0*x1 + x0

In [128]:
LiEG=lie("G2"); LiEG

G2

In [129]:
ad=lie.adjoint(LiEG); ad

1X[0,1]

In [130]:
lie.sym_tensor(4,ad,LiEG)

     1X[0,0] +2X[0,2] +1X[0,4] +1X[1,1] +2X[2,0] +1X[2,2] +1X[3,1] +
     1X[4,0]

### 5) LiE restriction matrices from Sage branching rules

From a SageMath branching rule we can extract how the fundamental weights of $\mathfrak{g}$ restrict to weights of $\mathfrak{h}$. This is precisely the information contained in the LiE restriction matrices. So if we don't know how to build a restriction matrix in LiE, we can see if SageMath has the branching rule.

In [131]:
def resmatrix(G,H,b):
    rkG=G.rank()
    fw=G.fundamental_weights()
    HWR=WeightRing(H)
    return [[int(x) for x in str(HWR(list(b(fw[i])))).split("(")[1].removesuffix(")").split(",")]\
               for i in range(1,rkG+1)]

Let us consider the space $\mathrm{SO}(mn)/(\mathrm{SO}(m)\times\mathrm{SO}(n))$, defined by the tensor product of standard representations
$$\mathrm{SO}(m)\times\mathrm{SO}(n)\to\mathrm{SO}(\mathbb{R}^m\otimes\mathbb{R}^n),\qquad (A,B)(v\otimes w)=Av\otimes Bw.$$
SageMath has the branching rule "tensor" for this.

In [132]:
def family(m,n):
    assert 3<=m and m<=n,"Make sure that 3 <= m <= n."
    bB1=branching_rule("B1","A1","isomorphic")
    bD2=branching_rule("D2","A1xA1","isomorphic")
    #lots of special cases due to SO(3) and SO(4). Sage is forgiving with the labels "B1" and "D2", but LiE isn't.
    if m==3:
        if n==3:
            Hstr="A1xA1"
            Gstr="B4"
            b=branching_rule("B4","B1xB1","tensor")*branching_rule("B1xB1","A1xA1",[bB1,bB1])
        elif n==4:
            Hstr="A1xA1xA1"
            Gstr="D6"
            b=branching_rule("D6","B1xD2","tensor")*branching_rule("B1xD2","A1xA1xA1",[bB1,bD2])
        elif n%2==0:
            Hstr="A1xD"+str(n/2)
            Gstr="D"+str(3*n/2)
            bId=branching_rule("D"+str(n/2),"D"+str(n/2),"identity")
            b=branching_rule(Gstr,"B1xD"+str(n/2),"tensor")*branching_rule("B1xD"+str(n/2),Hstr,[bB1,bId])
        else:
            Hstr="A1xB"+str((n-1)/2)
            Gstr="B"+str(3*(n-1)/2+1)
            bId=branching_rule("B"+str((n-1)/2),"B"+str((n-1)/2),"identity")
            b=branching_rule(Gstr,"B1xB"+str((n-1)/2),"tensor")*branching_rule("B1xB"+str((n-1)/2),Hstr,[bB1,bId])
    elif m==4:
        if n==4:
            Hstr="A1xA1xA1xA1"
            Gstr="D8"
            b=branching_rule("D8","D2xD2","tensor")*branching_rule("D2xD2","A1xA1xA1xA1",[bD2,bD2])
        elif n%2==0:
            Hstr="A1xA1xD"+str(n/2)
            Gstr="D"+str(2*n)
            bId=branching_rule("D"+str(n/2),"D"+str(n/2),"identity")
            b=branching_rule(Gstr,"D2xD"+str(n/2),"tensor")*branching_rule("D2xD"+str(n/2),Hstr,[bD2,bId])
        else:
            Hstr="A1xA1xB"+str((n-1)/2)
            Gstr="D"+str(2*n)
            bId=branching_rule("B"+str((n-1)/2),"B"+str((n-1)/2),"identity")
            b=branching_rule(Gstr,"D2xB"+str((n-1)/2),"tensor")*branching_rule("D2xB"+str((n-1)/2),Hstr,[bD2,bId])
    elif m%2==0:
        if n%2==0:
            Hstr="D"+str(m/2)+"xD"+str(n/2)
            Gstr="D"+str(m*n/2)
            b=branching_rule(Gstr,Hstr,"tensor")
        else:
            Hstr="D"+str(m/2)+"xB"+str((n-1)/2)
            Gstr="D"+str(m*n/2)
            b=branching_rule(Gstr,Hstr,"tensor")
    else:
        if n%2==0:
            Hstr="B"+str((m-1)/2)+"xD"+str(n/2)
            Gstr="D"+str(m*n/2)
            b=branching_rule(Gstr,Hstr,"tensor")
        else:
            Hstr="B"+str((m-1)/2)+"xB"+str((n-1)/2)
            Gstr="B"+str((m*n-1)/2)
            b=branching_rule(Gstr,Hstr,"tensor")
    G=WeylCharacterRing(Gstr,style="coroots")
    H=WeylCharacterRing(Hstr,style="coroots")
    return resmatrix(G,H,b)

In [134]:
Matrix(family(4,5))

[ 1  1  1  0]
[ 2  2  0  2]
[ 1  3  1  2]
[ 0  4  0  4]
[ 1  5  0  4]
[ 0  6  0  4]
[ 1  7 -1  4]
[ 2  8  0  2]
[ 1  4 -1  2]
[ 0  5  0  0]

### 6) The program LLBounds

This is a program to check the estimate $\Delta_{\mathrm{L}}\geq2E$ (on tt-tensors) on non-symmetric **normal homogeneous Einstein manifolds** $G/H$ with simple $G$, as classified by Krämer, Manturov, Wolf (isotropy irreducible), M. Wang and Ziller (isotropy reducible).

It relies on an expression of $\Delta_{\mathrm{L}}$ in terms of various Casimir operators, which are computed/estimated in SageMath, with the branching and tensoring happening inside LiE.

https://github.com/PSchwahn/LLBounds